### PR TITLE
 Generate CRD schema using github.com/kubernetes-sigs/controller-tools

### LINKED ISSUE #1687: Argo CD openapi validation spec is incomplete
**Describe the bug**

The openapi validation spec generated by https://github.com/kubernetes-sigs/controller-tools is incomplete. We should switch to controller-gen binary, to generate the CRD definition. It provides validation for all fields of the resource, which we were missing before, allowing a controller to be brought down by a single bad manifest.

**Version** v1.0.1

### LINKED ISSUE #1764: App Controller fails to save application due to spec validation error
**Describe the bug**
App Controller fails to save application due to spec validation error

**To Reproduce**
1. Create any app using kubectl which points to non-existing repo
2. Try to reconclie it
3. See error in controller logs

```

WARN[0013] Error updating application: Application.argoproj.io "testapp" is invalid: []: Invalid value: map[string]interface {}{"apiVersion":"argoproj.io/v1alpha1", "kind":"Application", "metadata":map[string]interface {}{"creationTimestamp":"2019-06-17T20:58:20Z", "generation":1, "name":"testapp", "namespace":"default", "resourceVersion":"6096491", "selfLink":"", "uid":"a33fd4da-9142-11e9-8b0a-025000000001", "clusterName":""}, "spec":map[string]interface {}{"destination":map[string]interface {}{"namespace":"default", "server":"https://localhost:6443"}, "project":"default", "source":map[string]interface {}{"path":"blue-green", "repoURL":"https://github.com/argoproj/argocd-example-apps123", "targetRevision":"HEAD"}}, "status":map[string]interface {}{"observedAt":"2019-06-17T20:58:47Z", "reconciledAt":"2019-06-17T20:58:47Z", "sync":map[string]interface {}{"comparedTo":map[string]interface {}{"destination":map[string]interface {}{"namespace":"default", "server":"https://localhost:6443"}, "source":map[string]interface {}{"path":"blue-green", "repoURL":"https://github.com/argoproj/argocd-example-apps123", "targetRevision":"HEAD"}}, "status":"Unknown"}, "conditions":[]interface {}{map[string]interface {}{"type":"ComparisonError", "message":"rpc error: code = Unknown desc = authentication required"}}, "health":map[string]interface {}{"status":"Healthy"}}}: validation failure list:
status.sync.revision in body is required  application=testapp
```

### PR DESCRIPTION
**Pull request by alexmt**
Closes #1764

### CODE REVIEW

File: ui/src/app/shared/services/applications-service.ts [hunk 1]
```diff
@@ -167,16 +168,19 @@ export class ApplicationsService {
     }
 
     private parseAppFields(data: any): models.Application {
-        const app = data as models.Application;
```
Comment (alexec) [inline ui/src/app/shared/services/applications-service.ts]: accidental?

**Review by alexec — APPROVED** [review]

**Comment by alexec** [comment]
I trust you on this. You need to fix your build errors :)

**Comment by alexec** [comment]
You've this error:

```
undefined: strings.ReplaceAll
```

That's because Golang 1.11 does not have that func. Do we need to upgrade :)

File: hack/gen-crd-spec/main.go [hunk 2]
```diff
@@ -0,0 +1,121 @@
...
+	objs, err := kube.SplitYAML(string(crdYamlBytes))
+	checkErr(err)
+	crds := make(map[string]*extensionsobj.CustomResourceDefinition)
+	for i := range objs {
+		un := objs[i]
+
+		// We need to completely remove validation of problematic fields such as creationTimestamp,
+		// which get marshalled to `null`, but are typed as as a `string` during Open API validation
+		removeValidataion(un, "metadata.creationTimestamp")
+		removeValidataion(un, "spec.template.metadata.creationTimestamp")
```
Comment (jessesuen) [inline hack/gen-crd-spec/main.go]: We don't need `spec.template.metadata.creationTimestamp`

**Review by jessesuen — APPROVED** [review]

File: hack/gen-crd-spec/main.go (hunk 2 above)
Comment (alexmt) [reply hack/gen-crd-spec/main.go]: done

**Pull request merged**